internal/webhook/workloadmutation/webhookhandler.go

```diff
@@ -0,0 +1,98 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package workloadmutation contains the webhook that injects annotations into daemon-sets, deployments and stateful-sets.
+package workloadmutation
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"net/http"
+
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/aws/amazon-cloudwatch-agent-operator/internal/config"
+	"github.com/aws/amazon-cloudwatch-agent-operator/pkg/instrumentation/auto"
+)
+
+// +kubebuilder:webhook:path=/mutate-v1-workload,mutating=true,failurePolicy=ignore,groups="apps",resources=daemonsets;deployments;statefulsets,verbs=create;update,versions=v1,name=mworkload.kb.io,sideEffects=none,admissionReviewVersions=v1
+// +kubebuilder:rbac:groups="apps",resources=daemonsets;deployments;statefulsets,verbs=get;list;watch
+
+var _ WebhookHandler = (*workloadMutationWebhook)(nil)
+
+// WebhookHandler is a webhook handler that analyzes new daemon-sets and injects appropriate annotations into it.
+type WebhookHandler interface {
+	admission.Handler
+}
+
+// the implementation.
+type workloadMutationWebhook struct {
+	client             client.Client
+	decoder            *admission.Decoder
+	logger             logr.Logger
+	config             config.Config
+	annotationMutation *auto.AnnotationMutators
+}
+
+// NewWebhookHandler creates a new WorkloadWebhookHandler.
+func NewWebhookHandler(cfg config.Config, logger logr.Logger, decoder *admission.Decoder, cl client.Client, annotationMutation *auto.AnnotationMutators) WebhookHandler {
+	return &workloadMutationWebhook{
+		config:             cfg,
+		decoder:            decoder,
+		logger:             logger,
+		client:             cl,
+		annotationMutation: annotationMutation,
+	}
+}
+
+func (p *workloadMutationWebhook) Handle(ctx context.Context, req admission.Request) admission.Response {
+	if p.annotationMutation == nil {
+		return admission.Errored(http.StatusBadRequest, errors.New("failed to unmarshal annotation config"))
```

Comment:
Not a blocker, but can we manually test for this scenario? Given our ignore failurePolicy, want to confirm we still proceed with the creation of the resource.

Reply:
To be fair, this is the same behavior as the podmutation webhook if it fails to decode the pod.

Instead of doing this check here, can we just change main to not create the webhook? If AnnotationMutators is nil, it means the config is invalid or the auto annotations were disabled.

Reply:
I am going to move this logic to main.go since this block workload creation. I do have the option to allow it but I think moving the logic to main makes mroe sense